### PR TITLE
Adding connection-backed Kubernetes Secrets to KubernetesPodOperator

### LINKED ISSUE #28086: Add support to KPO to manage secrets passing Connection-derived credentials
### Description

Add an option to automatically create, map, and manage secrets from Airflow Connection to a Kubernetes Secret for Kubernetes Pod Operator. 

### Use case/motivation

One of the important use-cases for Kubernetes Pod Operator is to be able to use credentials passed from Airflow Connections to the POD running the workload. 

Some of the users pass the credentials using environment variables, but this is inherently insecure, because environment variables passed to the Pod might be displayed and accessed using various mechanisms and certain scenarios (like failing to create Pod) might reveal it in various logs which are not protected by secret masker. 

This is possible even now, but it requires custom Kubernetes Pod Operator which would grab the credential, store it in a secret and mount the secret to the Pod via pod template and delete the secrets after the Pod completes.

This has some drawbacks - it is impossible to attach secret's lifecycle to the Pod lifecycle so managing such secrets (especially deleting them in failure scenarios) has to be done externally (for example by a cleanup script).

There is a "complex-ish" possibilty of managing those secrets by the operator itself. The best idea is to assign a fixed name to the secret (either connection or dag/task based), deleting them and handling potential race scenarios where "deletion" of the secret from one Pod completing would race with creating the secret while starting another Pod.

It would be great to implement such a feature for KPO. 

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by SamWheating**
closes: https://github.com/apache/airflow/issues/28086

This is very much a first-pass / WIP but I wanted to get some early input on the design decisions.

#### tl;dr

This allows a user to define a connection-backed secret to be consumed by the KubernetesPodOperator. The KPO handles the creation of these secrets at execution time and cleans them up after the fact.

A `KubernetesConnectionSecret` is almost identical to a regular `Secret`, except for the removal of the `secret` param and the addition of the `conn_id` param. A secret name is generated at run-time and added to the secret object, after which it can be used as a regular `Secret`.

#### Example:

```python
from airflow.providers.cncf.kubernetes.secret import KubernetesConnectionSecret
from airflow.providers.cncf.kubernetes.operators.pod import KubernetesPodOperator
from airflow.models import DAG
from datetime import datetime

# this will mount the password from the `my_conn` connection at `/etc/secret_password`
connection_secret = KubernetesConnectionSecret("volume", "/etc/secret_password", "my_conn", key="password")

with DAG(
    dag_id="kube_pod_with_connection_secrets",
    start_date=datetime(2021, 1, 1),
) as dag:
    k = KubernetesPodOperator(
        namespace="default",
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo", "10"],
        connection_secrets=[connection_secret],
        task_id="task",
    )
```

#### Open Questions:

 - How should we map a connection into the secret object? Currently I'm just writing out all of the fields to a dict.

#### TODO (before merging):

 - [ ] Additional test coverage in kubernetes_tests
 - [ ] Add documentation around this feature

### CODE REVIEW

**Comment by potiuk** [comment]
Very good idea.

cc: @eladkal - I believe you've been doing something similar in the past?

**Comment by eladkal** [comment]
> cc: @eladkal - I believe you've been doing something similar in the past?

Yes my idea was a bit different.
I was after:

```
    k = KubernetesPodOperator(
        namespace="default",
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo", "10"],
        connection_secrets=['postgres_default'],
        task_id="task",
    )
```

Where `connection_secrets` are Airflow connections.
the code read the value from db, encode it in base64 and user decode it with the code running in the pod.

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 1]
```diff
@@ -706,10 +715,51 @@ def post_complete_action(self, *, pod, remote_pod, **kwargs):
             remote_pod=remote_pod,
         )
 
+    def setup_connection_secrets(self, context):
+        """Create Kubernetes secrets based on Airflow connections."""
+        for s in self.connection_secrets:
+            name = f"{context['dag'].dag_id}-{context['task_instance'].task_id}-{_rand_str(8)}"
```
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/operators/pod.py:721]: This will not work as dag_id and task_id can contain chars not allowed in K8s.

You will need to handle it like:
    https://github.com/apache/airflow/blob/f971ba2f2f9703d0e1954e52aaded52a83c2f844/airflow/providers/cncf/kubernetes/operators/pod.py#L132

**Comment by SamWheating** [comment]
> the code read the value from db, encode it in base64 and store it as K8s secret (with unique suffix) then init_container can reference it. user decode it with the code running in the pod. When pod terminate the operator remove the secret created.

Interesting approach - in this approach I was trying to stay aligned with the existing `Secret` class and usage patterns, whereas this would create a largely different (albeit more flexible) avenue for defining and mounting secrets. Do you have any especially strong preferences or objections to the proposed implementation?

**Comment by hussein-awala** [comment]
> Yes my idea was a bit different.
I was after:
    k = KubernetesPodOperator(
        namespace="default",
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo", "10"],
        connection_secrets=['postgres_default'],
        task_id="task",
    )

That is what I do in my custom KPO operator, where it takes a list of (connection_id, env_var_prefix) as argument, and it creates a single secret in runtime and mounts the secret using `deploy_type=env`, for the secrets names, I use `<env_var_prefix>_LOGIN`, `<env_var_prefix>_PASSWORD`, ... and all these fields are available in my pod as environment variables. At the end of the execution this secret is deleted using the cleanup or on_kill method.

**Comment by eladkal** [comment]
> > Yes my idea was a bit different.
> > I was after:
> > k = KubernetesPodOperator(
> > namespace="default",
> > image="ubuntu:16.04",
> > cmds=["bash", "-cx"],
> > arguments=["echo", "10"],
> > connection_secrets=['postgres_default'],
> > task_id="task",
> > )
> 
> That is what I do in my custom KPO operator, where it takes a list of (connection_id, env_var_prefix) as argument, and it creates a single secret in runtime and mounts the secret using `deploy_type=env`, for the secrets names, I use `<env_var_prefix>_LOGIN`, `<env_var_prefix>_PASSWORD`, ... and all these fields are available in my pod as environment variables. At the end of the execution this secret is deleted using the cleanup or on_kill method.

Almost exactly like me.
I did have issues with cleanup but I guess because I deployed on celery. If worker died the cleanup code was never executed so I had to do manual cleanup from time to time.

So if we both do the same we never talked about this before.. i guess that is a good approach :)

**Comment by SamWheating** [comment]
> My motivation was that it is far more likely that when you want to inject a connection to K8s pod it would be one already defined in Airflow connections and being used by other operators rather than one stored elsewhere and you want to just bake it into the task. Also if that was not the case all you need is just to define a new connection. Basicly anything Airflow knows will also be accesable from K8s pod.

Agreed - the same is true of this implementation.

> I think it's also easier to explain to users - just tell us what is the conn_id you want and we will make sure it's available in the pod. When referencing it just decode the attributes.

I think that this actually adds some complexity for users, as they have to understand two largely different mechanisms for mounting secrets. I also feel like it removes a lot of configurability with the existing secrets, such as mounting in env vs. a volume, which keys to mount etc. Since we already have all of this in place in this operator (with `secrets=[]`), why not re-use it?

> it takes a list of (connection_id, env_var_prefix) as argument, and it creates a single secret in runtime and mounts the secret using deploy_type=env, for the secrets names, I use <env_var_prefix>_LOGIN, <env_var_prefix>_PASSWORD

I have similar thoughts about this approach - it introduces more hidden functionality and requires users to be familiar with the implementation details and how they differ from the existing `secrets: list[Secret]` implementation.

**Comment by potiuk** [comment]
Looking at the discussion... just a thought (and I think it comes very closely to a thought that @bolkedebruin had about making standalone tasks without the task havint to reach out to the DB of airflow).

Why don't we make those secrets to be COMPATIBLE with the way airflow tasks use it ?

I believe K8S secrets can be mapped to environment variables.... 

Why don't we map the connection secrets to Kubernetes secrets that are in the format of environment variables that follow Airflow's connection defiinitions for env variables?

If we can map such connections to AIRFLOW__CONNECTION_XXXX=<encoded url for connection> - then we can simply run any airflow-compatible Hook as K8S Pod and it will automatically retrieve connnection from the env variable. So basically we could run any Airflow Hook inside the K8S Pod run via K8SPodOperator and that Hook would **just work** (without accessing the DB of airflow - just reading the env variable coming from the mounted secret.

Unless it was already going to work like that and I misunderstood it ..

**Comment by jedcunningham** [comment]
It's not a bad idea to make it use the Airflow env var format just for consistency, but not for the "ease of use" with hooks like you are thinking of. KPO typically doesn't run real "Airflow" code, and you need a whole lot more than just some connections to make that work (really need all config, dags, etc) - it needs to look more like a KE worker, which is non-trivial with KPO.

My 2c, making it feel more like other operators where you provide the connection id, and KPO adds the normal `AIRFLOW_CONN_{CONN_ID}` env var is a better authoring experience than trying to mirror the existing secret interface.

**Comment by jedcunningham** [comment]
I'll add, the orphaned secrets problem does worry me a bit. Doesn't feel great that we can leave stuff in secrets sitting around potentially forever. I get why folks have chosen to build it in custom wrappers, but it worries me making it a product thing 🤷‍♂️.

And I don't have a good mitigation or alternative here. We can't even have the scheduler do occasional cleanup, since we can go create thing things in many k8s contexts too.

**Comment by hussein-awala** [comment]
> I'll add, the orphaned secrets problem does worry me a bit. Doesn't feel great that we can leave stuff in secrets sitting around potentially forever. I get why folks have chosen to build it in custom wrappers, but it worries me making it a product thing 🤷‍♂️.

I had not this problem for different reasons:
- In the past I had a big grace period for celery executors, so the worker was rarely forcefully killed (they were running on normal machine and not SPOT)
- When the task restarts, `get_or_create_pod` finds the same pod instead of creating a new one, and I could do the same thing for the secrets
- When I migrated to deferrable operators, the problem was completely solved, because the triggerer is fault tolerant, it can resume executing the trigger when it restart, and the task is not running when it's deferred
- For the cleanup, there were different solutions, one of them is patching the `ownerReferences` for the secret after creating the pod, in this case the secret will be automatically killed when we kill the pod.

So the possibility of having these orphaned secrets is too small.

**Comment by potiuk** [comment]
> * or the cleanup, there were different solutions, one of them is patching the `ownerReferences` for the secret after creating the pod, in this case the secret will be automatically killed when we kill the pod.

That one looks pretty cool actually

**Comment by SamWheating** [comment]
> I get why folks have chosen to build it in custom wrappers, but it worries me making it a product thing 🤷‍♂️.

Agreed, especially given that kubernetes secrets are stored unencrypted in etcd and often easily accessible to other workloads and users 😬

**Comment by eladkal** [comment]
> I think that this actually adds some complexity for users, as they have to understand two largely different mechanisms for mounting secrets. I also feel like it removes a lot of configurability with the existing secrets, such as mounting in env vs. a volume, mounting a subset of fields etc. Since we already have all of this in place in this operator (with secrets: list[Secret]), why not re-use it?

Any design choice you take has advantages and disadvantages.
I am a strong advocate for allowing referencing Airflow conn_id. From functionality point of view it makes no sense that we have a connection details stored in Airflow but for K8s pod users will have to redefine it somewhere else.

The how to make this work in K8s is another topic.


> It indeed crossed my mind too, and it is quite a bit worrying, I quite agree. I know @eladkal 's team approach was exactly this, they had jobs that cleaned old secrets. Unfortunately K8S has no auto-cleaninig of secrets. It would be a great feature to have. But we don't....

Yes but it wasn't that much of a pain. It's a simple script to run once a week and it takes less than 1 min to finish.

**Comment by potiuk** [comment]
> I am a strong advocate for allowing referencing Airflow conn_id. From functionality point of view it makes no sense that we have a connection details stored in Airflow but for K8s pod users will have to redefine it somewhere else.

I am with Elad here. That's why  https://github.com/apache/airflow/pull/33680#issuecomment-1692475222 comment. I think we can easily make such secrets 100% compatible with what Airflow's expectation is.

> Yes but it wasn't that much of a pain. It's a simple script to run once a week and it takes less than 1 min to finish.

See above:

>> or the cleanup, there were different solutions, one of them is patching the ownerReferences for the secret after creating the pod, in this case the secret will be automatically killed when we kill the pod.
> That one looks pretty cool actually

Seems that `ownerReference` might be the way to go - no cleanup necessary and the secret will get automatically deleted when the owner (i.e. our k8s POD) is gone. Not sure if that is possible to set (I am afraid we might have a chicken-egg problem, but maybe:

1) creating secret
2) creating pod
3) patching secret with ovnerReference of the pod created 

Is the way to go. The only small race is if our code gets killed between 2) and 3) but this can be either neglected or cleanup script mentioned by Elad could help to cope with the edge case.

**Comment by SamWheating** [comment]
Ok, I think we have a lot of good suggestions and sufficient consensus on how this feature should look. I will update the implementation accordingly and then open a new PR for discussion.

**Comment by rahulnbhandari** [comment]
> Ok, I think we have a lot of good suggestions and sufficient consensus on how this feature should look. I will update the implementation accordingly and then open a new PR for discussion.

I like this feature/ approach .

**Comment by potiuk** [comment]
Not stale. @SamWheating - are you going to maybe complete that one :) ?

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 2]
```diff
@@ -706,10 +715,51 @@ def post_complete_action(self, *, pod, remote_pod, **kwargs):
...
+            secret_body = {
+                "extra": conn.extra,
+                "host": conn.host,
+                "login": conn.login,
+                "password": conn.password,
+                "port": conn.port,
+                "schema": conn.schema,
+            }
+            # create secret
+            self.client.create_namespaced_secret(
```
Comment (NotYuki) [inline airflow/providers/cncf/kubernetes/operators/pod.py:734]: What would happen if we pass multiple Secret objects with the same conn_id but different keys to KPO? Is it accurate to say that, for each Secret, this code will generate a complete Kubernetes secret with six keys, mapping only one key from each secret to a specific environment variable in the pod?

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 3]
```diff
@@ -706,10 +715,51 @@ def post_complete_action(self, *, pod, remote_pod, **kwargs):
...
+            # use it alongside the other KubernetesPodOperator secrets.
+            self.secrets.append(s)
+
+    def cleanup_connection_secrets(self):
+        """Delete any ephemeral Kubernetes secrets created by connection_secrets."""
+        for s in self.secrets:
+            if isinstance(s, KubernetesConnectionSecret):
+                self.log.info("Attempting to delete connection-backed secret: %s", s.secret)
+                try:
+                    self.client.delete_namespaced_secret(self.namespace, s.secret)
```
Comment (NotYuki) [inline airflow/providers/cncf/kubernetes/operators/pod.py:755]: It looks like you need to swap arguments in delete_namespaced_secret() call.
Signature of the method:
`delete_namespaced_secret(self, name, namespace, **kwargs)`

**Comment by SamWheating** [comment]
I really appreciate the discussion here, but its apparent that I don't have the capacity at the moment to get the promised rewrite done.

I'm gonna close this PR and un-assign myself from the ticket, hopefully someone else can pick this up and complete it 🤞
